capability.go

```diff
@@ -572,6 +582,18 @@ func (c Client) Release() {
 	h.Shutdown()
 }
 
+func (c Client) EncodeAsPtr(seg *Segment) Ptr {
+	capId := seg.Message().AddCap(c)
+	return NewInterface(seg, capId).ToPtr()
+}
+
+func (c *Client) DecodeFromPtr(p Ptr) {
```

Comment:
A side-effect of using pointer receivers is that we will not be able to assign concrete types to `TypeParam`.  It seems like most functions in this library try to avoid returning pointers, so I'm worried this might make composition a bit awkward.

The radical solution would be to deal with `*Ptr`, `*Struct`, `*List` and `*Client` everywhere, but I suspect there's a good reason we avoided that in the first place.

Reply:
Ack, you're right. I suppose we could go back to the solution I had before, where the interface has a type parameter that's supposed to match the receiver, and you have to write `T TypeParam[T]` everywhere, but then decode can take the raw type, rather than a pointer...

I wish I had a way to write generic functions like:

```go
func Foo[*T TypeParam](...) T
```

i.e. specify a constraint that `*T` implements TypeParam, but still have T be the type parameter...

You *can* return a pointer by doing:

```go
type TPPtr[A any] interface {
	TypeParam
	~*A
}

func Foo[A any, PtrA TPPtr[A]](p Ptr) A {
	var a A
	b := PtrA(&a)
	b.DecodeFromPtr(p)
	return a
}

// ...

Foo[A, *A](p)
```

...but that's pretty horrible...

Reply:
Ok, I did a second pass where I tweaked the interface to go back to returning the value, which gets rid of the pointer receiver. A bit more convoluted, but will make call sites better, and it's not really used except by generated code.

Reply:
😚👌